# História das Olimpíadas  
_(créditos ao prof. Rafael Moreira)_

Após um ano de atraso por conta da pandemia de Covid-19, as atenções do mundo todo se voltaram para Tóquio, no Japão, para acompanhar mais uma edição das Olimpíadas.

No Brasil não foi diferente, e muitos se uniram para torcer por nossos atletas em diferentes competições, tanto em esportes onde o Brasil já possui tradição quanto em novos esportes.

Vamos aproveitar o clima para estudar um pouco das Olimpíadas! Utilizaremos um _dataset_ com 120 anos de dados históricos das Olimpíadas, cobrindo desde os jogos de Atenas 1896 até Rio 2016. 

Faça o download do _dataset_ em https://www.kaggle.com/heesoo37/120-years-of-olympic-history-athletes-and-results e carregue o arquivo ```athlete_events.csv``` para um DataFrame utilizando Pandas. Aproveite para explorar seu DataFrame e se familiarizar com a sua estrutura. 

OBS: Fique à vontade para acrescentar mais células Python conforme necessário em qualquer etapa do exercício.

# <span style='color:Tomato'>NOTA PARA O PROFESSOR</span>

Títulos com a cor em <span style='color:Tomato'>vermelho</span> significa que não fiquei totalmente satisfeito com a lógica do código, ele dá o resultado esperado, porém feito com certas "gambiarras" ou de modo que não fique escalável caso os valores do data frame sejam modificados.

##  <span style='color:LightSeaGreen'>Sumário</span>
0. [Preparando o Ambiente](#env)
1. [O Brasil nas Olimpíadas](#pt1)  
    -[Atletas brasileiros](#pt1.1)  
    -[Medalhistas](#pt1.2)  
    -[Verão vs Inverno](#pt1.3)  
    -[Atletas do Brasil](#pt1.4)  
2. [O mundo nos jogos de verão](#pt2)
3. [Brasil vs Mundo](#pt3)

## 0. Preparando o Ambiente <span id='env'>

In [361]:
import pandas as pd
import re
from bs4 import BeautifulSoup as bs
from urllib.request import urlopen

# solves urlopen error [SSL: CERTIFICATE_VERIFY_FAILED]
import ssl
ssl._create_default_https_context = ssl._create_unverified_context

df = pd.read_csv('./dados/athlete_events.csv')
df

ID                      Name Sex   Age  Height  Weight  \
0            1                 A Dijiang   M  24.0   180.0    80.0   
1            2                  A Lamusi   M  23.0   170.0    60.0   
2            3       Gunnar Nielsen Aaby   M  24.0     NaN     NaN   
3            4      Edgar Lindenau Aabye   M  34.0     NaN     NaN   
4            5  Christine Jacoba Aaftink   F  21.0   185.0    82.0   
...        ...                       ...  ..   ...     ...     ...   
271111  135569                Andrzej ya   M  29.0   179.0    89.0   
271112  135570                  Piotr ya   M  27.0   176.0    59.0   
271113  135570                  Piotr ya   M  27.0   176.0    59.0   
271114  135571        Tomasz Ireneusz ya   M  30.0   185.0    96.0   
271115  135571        Tomasz Ireneusz ya   M  34.0   185.0    96.0   

                  Team  NOC        Games  Year  Season            City  \
0                China  CHN  1992 Summer  1992  Summer       Barcelona   
1                China  CHN  2012 Summer  2012  Summer          London   
2              Denmark  DEN  1920 Summer  1920  Summer       Antwerpen   
3       Denmark/Sweden  DEN  1900 Summer  1900  Summer           Paris   
4          Netherlands  NED  1988 Winter  1988  Winter         Calgary   
...                ...  ...          ...   ...     ...             ...   
271111        Poland-1  POL  1976 Winter  1976  Winter       Innsbruck   
271112          Poland  POL  2014 Winter  2014  Winter           Sochi   
271113          Poland  POL  2014 Winter  2014  Winter           Sochi   
271114          Poland  POL  1998 Winter  1998  Winter          Nagano   
271115          Poland  POL  2002 Winter  2002  Winter  Salt Lake City   

                Sport                                     Event Medal  
0          Basketball               Basketball Men's Basketball   NaN  
1                Judo              Judo Men's Extra-Lightweight   NaN  
2            Football                   Football Men's Football   NaN  
3          Tug-Of-War               Tug-Of-War Men's Tug-Of-War  Gold  
4       Speed Skating          Speed Skating Women's 500 metres   NaN  
...               ...                                       ...   ...  
271111           Luge                Luge Mixed (Men)'s Doubles   NaN  
271112    Ski Jumping  Ski Jumping Men's Large Hill, Individual   NaN  
271113    Ski Jumping        Ski Jumping Men's Large Hill, Team   NaN  
271114      Bobsleigh                      Bobsleigh Men's Four   NaN  
271115      Bobsleigh                      Bobsleigh Men's Four   NaN  

[271116 rows x 15 columns]

## <span id='pt1' style='color:ForestGreen'><b>1. O Brasil nas Olimpíadas </span>

### <span id='pt1.1' style='color:Gold'> Atletas brasileiros </span>

Vamos começar estudando o desempenho do nossos próprio país. Gere um DataFrame novo contendo apenas as informações sobre atletas brasileiros.

In [362]:
df[df['Team']=='Brazi-1']
df[df['Name'].str.contains(r'^Ana Paula.*')]
year_2004_df = df[df['Year']==2004].copy()
print(year_2004_df[year_2004_df['Team'].str.contains(r'^Brazil.*')]['Team'].value_counts())
year_2004_df[year_2004_df['Team'].str.contains(r'^Brazil-.*')].sort_values('Medal')

Brazil      310
Brazil-2      4
Brazil-1      4
Name: Team, dtype: int64


ID                                              Name Sex   Age  \
63345    32468                   Emanuel Fernando Scheffler Rego   M  31.0   
199962  100430                         Ricardo Alex Costa Santos   M  29.0   
17628     9394                              Adriana Brando Behar   F  35.0   
217980  109492                           Shelda Kelly Bruno Bede   F  31.0   
6448      3615  Ana Paula Rodrigues Connelly (-Rodrigues Henkel)   F  32.0   
19120    10125                                  Benjamin Insfran   M  32.0   
149213   74846                      Mrcio Henrique Barroso Arajo   M  30.0   
209880  105385                   Sandra Tavares Pires Nascimento   F  31.0   

        Height  Weight      Team  NOC        Games  Year  Season    City  \
63345    190.0    80.0  Brazil-1  BRA  2004 Summer  2004  Summer  Athina   
199962   200.0   102.0  Brazil-1  BRA  2004 Summer  2004  Summer  Athina   
17628    180.0    64.0  Brazil-1  BRA  2004 Summer  2004  Summer  Athina   
217980   165.0    59.0  Brazil-1  BRA  2004 Summer  2004  Summer  Athina   
6448     183.0    68.0  Brazil-2  BRA  2004 Summer  2004  Summer  Athina   
19120    196.0    97.0  Brazil-2  BRA  2004 Summer  2004  Summer  Athina   
149213   192.0    89.0  Brazil-2  BRA  2004 Summer  2004  Summer  Athina   
209880   174.0    64.0  Brazil-2  BRA  2004 Summer  2004  Summer  Athina   

                   Sport                                      Event   Medal  
63345   Beach Volleyball    Beach Volleyball Men's Beach Volleyball    Gold  
199962  Beach Volleyball    Beach Volleyball Men's Beach Volleyball    Gold  
17628   Beach Volleyball  Beach Volleyball Women's Beach Volleyball  Silver  
217980  Beach Volleyball  Beach Volleyball Women's Beach Volleyball  Silver  
6448    Beach Volleyball  Beach Volleyball Women's Beach Volleyball     NaN  
19120   Beach Volleyball    Beach Volleyball Men's Beach Volleyball     NaN  
149213  Beach Volleyball    Beach Volleyball Men's Beach Volleyball     NaN  
209880  Beach Volleyball  Beach Volleyball Women's Beach Volleyball     NaN

In [363]:
df_brazil = df[df['Team'].str.contains(r'^Brazil.*')].copy()
df_brazil.reset_index(drop=True, inplace=True)
df_brazil.head()

ID                        Name Sex   Age  Height  Weight    Team  NOC  \
0  386             Alexandre Abeid   M  22.0   194.0    92.0  Brazil  BRA   
1  386             Alexandre Abeid   M  26.0   194.0    92.0  Brazil  BRA   
2  388  Abel Carlos da Silva Braga   M  19.0   190.0    73.0  Brazil  BRA   
3  451         Diana Monteiro Abla   F  21.0   175.0    75.0  Brazil  BRA   
4  565        Glauclio Serro Abreu   M  26.0   185.0    75.0  Brazil  BRA   

         Games  Year  Season            City       Sport  \
0  1972 Summer  1972  Summer          Munich  Volleyball   
1  1976 Summer  1976  Summer        Montreal  Volleyball   
2  1972 Summer  1972  Summer          Munich    Football   
3  2016 Summer  2016  Summer  Rio de Janeiro  Water Polo   
4  2004 Summer  2004  Summer          Athina      Boxing   

                           Event Medal  
0    Volleyball Men's Volleyball   NaN  
1    Volleyball Men's Volleyball   NaN  
2        Football Men's Football   NaN  
3  Water Polo Women's Water Polo   NaN  
4      Boxing Men's Middleweight   NaN

### <span id='pt1.2' style='color:Gold'> Medalhistas </span>

Vamos focar um pouco nos casos de sucesso do Brasil. Use o seu DataFrame anterior para filtrar apenas informações sobre **medalhistas** brasileiros. 

**DICA:** observe como a coluna ```Medal``` é representada quando o atleta não ganhou medalha.

In [364]:
df_brazil_medals = df_brazil[df_brazil['Medal'].notna()].sort_values('Medal').reset_index(drop=True)
df_brazil_medals.head()

ID                             Name Sex   Age  Height  Weight      Team  \
0  55088            Andr Bier Johannpeter   M  37.0   178.0    78.0    Brazil   
1  97270                   Nlson Prudncio   M  28.0   182.0    71.0    Brazil   
2  32468  Emanuel Fernando Scheffler Rego   M  35.0   190.0    80.0  Brazil-2   
3  96701                      Bruno Prada   M  40.0   185.0   110.0    Brazil   
4  93746          Rodrigo de Paula Pessoa   M  27.0   177.0    67.0    Brazil   

   NOC        Games  Year  Season     City             Sport  \
0  BRA  2000 Summer  2000  Summer   Sydney     Equestrianism   
1  BRA  1972 Summer  1972  Summer   Munich         Athletics   
2  BRA  2008 Summer  2008  Summer  Beijing  Beach Volleyball   
3  BRA  2012 Summer  2012  Summer   London           Sailing   
4  BRA  2000 Summer  2000  Summer   Sydney     Equestrianism   

                                     Event   Medal  
0        Equestrianism Mixed Jumping, Team  Bronze  
1              Athletics Men's Triple Jump  Bronze  
2  Beach Volleyball Men's Beach Volleyball  Bronze  
3        Sailing Men's Two Person Keelboat  Bronze  
4        Equestrianism Mixed Jumping, Team  Bronze

In [365]:
df_brazil_medals['Medal'].value_counts()

Bronze    191
Silver    175
Gold      109
Name: Medal, dtype: int64

### <span id='pt1.3' style='color:Gold'> Verão vs Inverno </span>

Você deve ter notado que temos duas categorias distintas de jogos olímpicos, representados pela estação: temos os jogos de verão e os jogos de inverno, que ocorrem de maneira intercalada.

Agora que já conhecemos os medalhistas brasileiros, resposta: quantos atletas brasileiros receberam medalha nos jogos de verão e quantos receberam nos jogos de inverno?

In [366]:
df_brazil_medals['Season'].value_counts()

Summer    475
Name: Season, dtype: int64

In [367]:
df_brazil[df_brazil['Season']=='Winter']['Medal'].value_counts()

Series([], Name: Medal, dtype: int64)

<span style='color:LightGreen'>Todas as <b>475</b> medalhas são dos jogos de verão, não há medalhas em jogos de inverno.</span>

Os jogos de verão são bem mais populares do que os jogos de inverno no Brasil. Portanto, deste ponto em diante iremos focar apenas nos jogos de verão. Descarte de seu DataFrame os dados dos jogos de inverno.



In [368]:
df_brazil_medals_summer = df_brazil_medals[df_brazil_medals['Season']=='Summer']

### <span id='pt1.4' style='color:Gold'> Atletas do Brasil </span>

In [369]:
####

In [370]:
brazilian_mean_height = df_brazil_medals_summer['Height'].mean()
brazilian_mean_weight = df_brazil_medals_summer['Weight'].mean()
print(f'Altura média dos atletas: {round(brazilian_mean_height,2)} cm')
print(f'Peso médio dos atletas: {round(brazilian_mean_weight,2)} Kg')

Altura média dos atletas: 182.49 cm
Peso médio dos atletas: 76.71 Kg


#### <span id='pt1.4.2'></span>Imaginamos que diferentes esportes podem beneficiar diferentes tipos físicos, certo? Então refaça a análise anterior, mas obtendo os valores médios **por esporte**.

In [371]:
print('Altura média dos atletas por esporte:')
medalist_height_by_sport = df_brazil_medals_summer.groupby(by='Sport')['Height'].mean().apply(lambda value:round(value,2))
medalist_height_by_sport

Altura média dos atletas por esporte:


Sport
Athletics            181.00
Basketball           185.61
Beach Volleyball     184.88
Boxing               170.00
Canoeing             175.00
Equestrianism        179.67
Football             175.80
Gymnastics           162.75
Judo                 176.67
Modern Pentathlon    166.00
Sailing              181.59
Shooting             175.00
Swimming             189.11
Taekwondo            184.00
Volleyball           190.59
Name: Height, dtype: float64

In [372]:
print('Peso médio dos atletas por esporte:')
medalist_weight_by_sport = df_brazil_medals_summer.groupby(by='Sport')['Weight'].mean().apply(lambda value:round(value,2))
medalist_weight_by_sport

Peso médio dos atletas por esporte:


Sport
Athletics            74.58
Basketball           78.48
Beach Volleyball     78.16
Boxing               64.00
Canoeing             83.25
Equestrianism        75.00
Football             69.96
Gymnastics           63.75
Judo                 86.29
Modern Pentathlon    55.00
Sailing              80.41
Shooting             69.00
Swimming             81.56
Taekwondo            79.50
Volleyball           81.17
Name: Weight, dtype: float64

#### <span id='pt1.4.3'></span>Será que os dados acima influenciaram no interesse geral dos atletas pelo esporte ou realmente impactaram no desempenho deles? Podemos tentar descobrir se há algum tipo de correlação.

Você ainda possui o dataframe original contendo todos os atletas brasileiros, incluindo os sem medalha? Obtenha os valores médios de peso e altura por esporte daquele dataframe e compare-o com os dos medalhistas. Há alguma diferença significativa em algum esporte?

In [373]:
# Height analysis
# Create data frame with data from medalist vs all
all_height_by_sport = df_brazil.groupby(by='Sport')['Height'].mean().apply(lambda value:round(value,2))
all_weight_by_sport = df_brazil.groupby(by='Sport')['Weight'].mean().apply(lambda value:round(value,2))

In [374]:
# merge height data
medal_vs_all_height = pd.merge(medalist_height_by_sport, all_height_by_sport, \
            left_index=True, right_index=True,\
            suffixes=('_Medalist', '_All')).sort_index()
# add a column with the height difference between medalist and everyone else
medal_vs_all_height['Height_diff_percent'] = (medal_vs_all_height['Height_Medalist'] - medal_vs_all_height['Height_All'])\
                                    /medal_vs_all_height['Height_Medalist']*100
# merge weight data)
medal_vs_all_weight = pd.merge(medalist_weight_by_sport, all_weight_by_sport, \
            left_index=True, right_index=True,\
            suffixes=('_Medalist', '_All')).sort_index()
# add a column with the height difference between medalist and everyone else
medal_vs_all_weight['Weight_diff_percent'] = (medal_vs_all_weight['Weight_Medalist'] - medal_vs_all_weight['Weight_All'])\
                                    /medal_vs_all_weight['Weight_Medalist']*100

# visualizing both data
pd.concat([medal_vs_all_height,medal_vs_all_weight], axis=1)

Height_Medalist  Height_All  Height_diff_percent  \
Sport                                                                 
Athletics                   181.00      176.20             2.651934   
Basketball                  185.61      190.91            -2.855450   
Beach Volleyball            184.88      185.70            -0.443531   
Boxing                      170.00      171.99            -1.170588   
Canoeing                    175.00      177.79            -1.594286   
Equestrianism               179.67      177.43             1.246730   
Football                    175.80      173.37             1.382253   
Gymnastics                  162.75      157.27             3.367127   
Judo                        176.67      173.93             1.550914   
Modern Pentathlon           166.00      168.60            -1.566265   
Sailing                     181.59      179.59             1.101382   
Shooting                    175.00      174.04             0.548571   
Swimming                    189.11      181.18             4.193327   
Taekwondo                   184.00      173.77             5.559783   
Volleyball                  190.59      187.76             1.484863   

                   Weight_Medalist  Weight_All  Weight_diff_percent  
Sport                                                                
Athletics                    74.58       67.80             9.090909  
Basketball                   78.48       85.90            -9.454638  
Beach Volleyball             78.16       78.62            -0.588536  
Boxing                       64.00       64.11            -0.171875  
Canoeing                     83.25       77.40             7.027027  
Equestrianism                75.00       72.31             3.586667  
Football                     69.96       67.76             3.144654  
Gymnastics                   63.75       52.46            17.709804  
Judo                         86.29       80.29             6.953297  
Modern Pentathlon            55.00       60.67           -10.309091  
Sailing                      80.41       76.81             4.477055  
Shooting                     69.00       76.92           -11.478261  
Swimming                     81.56       74.59             8.545856  
Taekwondo                    79.50       63.62            19.974843  
Volleyball                   81.17       78.85             2.858199

In [375]:
# for height data, Taekwondo, Swimming, Gymnastics, Basketball have the most difference

heigh_diff = medal_vs_all_height['Height_diff_percent'].apply(lambda value: abs(value)).sort_values(ascending=False)
heigh_diff[heigh_diff > heigh_diff.describe()['75%']]

Sport
Taekwondo     5.559783
Swimming      4.193327
Gymnastics    3.367127
Basketball    2.855450
Name: Height_diff_percent, dtype: float64

In [376]:
# for weight data, Taekwondo, Gymnastics, Shooting and Modern Pentathlon have the most difference

weigh_diff = medal_vs_all_weight['Weight_diff_percent'].apply(lambda value: abs(value)).sort_values(ascending=False)
weigh_diff[weigh_diff > weigh_diff.describe()['75%']]

Sport
Taekwondo            19.974843
Gymnastics           17.709804
Shooting             11.478261
Modern Pentathlon    10.309091
Name: Weight_diff_percent, dtype: float64

#### <span id='pt1.4.4'></span>Existe um detalhe importante passando batido até agora em nossa análise: as categorias esportivas costumam ser divididas por gênero justamente por conta de diferenças físicas entre homens e mulheres que poderiam influenciar no desempenho. Compare a altura e peso médios de atletas brasileiros por esporte segmentado por sexo.

In [377]:
# verifying how many gender types
df_brazil['Sex'].value_counts()

M    2680
F    1148
Name: Sex, dtype: int64

In [378]:
# data frame with only male
df_brazil_male = df_brazil[df_brazil['Sex']=='M']
# data frame with only female
df_brazil_female = df_brazil[df_brazil['Sex']=='F']
# get height mean values by sport for males
brazil_male_height_by_sport = df_brazil_male.groupby(by='Sport')['Height'].mean().apply(lambda value: round(value,2))
# get height mean values by sport for females
brazil_female_height_by_sport = df_brazil_female.groupby(by='Sport')['Height'].mean().apply(lambda value: round(value,2))
# merge both series
male_vs_female_height = pd.concat([brazil_male_height_by_sport,brazil_female_height_by_sport], axis=1).dropna()
male_vs_female_height.columns = ['Male_Height','Female_Height']
# add a column with the height difference between male and female
male_vs_female_height['Height_diff_percent'] = round(abs((male_vs_female_height['Male_Height'] - male_vs_female_height['Female_Height'])\
                                    /male_vs_female_height['Male_Height']*100),2)

# same procedure for weight
brazil_male_weight_by_sport = df_brazil_male.groupby(by='Sport')['Weight'].mean().apply(lambda value: round(value,2))
brazil_female_weight_by_sport = df_brazil_female.groupby(by='Sport')['Weight'].mean().apply(lambda value: round(value,2))
male_vs_female_weight = pd.concat([brazil_male_weight_by_sport,brazil_female_weight_by_sport], axis=1).dropna()
male_vs_female_weight.columns = ['Male_Weight','Female_Weight']
male_vs_female_weight['Weight_diff_percent'] = round(abs((male_vs_female_weight['Male_Weight'] - male_vs_female_weight['Female_Weight'])\
                                    /male_vs_female_weight['Male_Weight']*100),2)

# visualize data
pd.concat([male_vs_female_height,male_vs_female_weight],axis=1)

Male_Height  Female_Height  Height_diff_percent  \
Sport                                                                   
Alpine Skiing              184.15         160.33                12.94   
Archery                    177.85         162.86                 8.43   
Athletics                  180.07         167.49                 6.99   
Badminton                  183.00         168.00                 8.20   
Basketball                 195.46         182.87                 6.44   
Beach Volleyball           194.54         176.48                 9.28   
Bobsleigh                  185.00         168.50                 8.92   
Boxing                     172.34         167.60                 2.75   
Canoeing                   178.55         169.50                 5.07   
Cross Country Skiing       178.75         167.25                 6.43   
Cycling                    176.41         164.31                 6.86   
Diving                     171.90         161.20                 6.22   
Equestrianism              178.17         169.00                 5.15   
Fencing                    181.20         171.83                 5.17   
Football                   177.26         166.62                 6.00   
Golf                       175.00         167.50                 4.29   
Gymnastics                 168.92         150.54                10.88   
Handball                   187.41         175.96                 6.11   
Judo                       177.10         167.12                 5.64   
Modern Pentathlon          169.64         165.75                 2.29   
Rowing                     181.40         172.33                 5.00   
Rugby Sevens               176.58         166.31                 5.82   
Sailing                    181.52         170.00                 6.35   
Shooting                   175.56         165.62                 5.66   
Swimming                   184.95         169.39                 8.41   
Table Tennis               174.00         163.78                 5.87   
Taekwondo                  176.00         171.86                 2.35   
Tennis                     188.00         169.78                 9.69   
Triathlon                  184.00         165.78                 9.90   
Volleyball                 192.31         181.44                 5.65   
Water Polo                 181.42         172.92                 4.69   
Weightlifting              171.78         161.50                 5.98   
Wrestling                  182.14         169.33                 7.03   

                      Male_Weight  Female_Weight  Weight_diff_percent  
Sport                                                                  
Alpine Skiing               85.15          52.33                38.54  
Archery                     78.15          59.71                23.60  
Athletics                   71.55          59.81                16.41  
Badminton                   78.00          70.00                10.26  
Basketball                  92.95          73.38                21.05  
Beach Volleyball            91.75          64.91                29.25  
Bobsleigh                   89.75          74.50                16.99  
Boxing                      64.03          65.00                 1.51  
Canoeing                    78.57          64.50                17.91  
Cross Country Skiing        74.50          53.75                27.85  
Cycling                     69.97          53.50                23.54  
Diving                      69.58          59.13                15.02  
Equestrianism               73.77          57.17                22.50  
Fencing                     76.57          61.58                19.58  
Football                    72.54          60.09                17.16  
Golf                        75.00          55.50                26.00  
Gymnastics                  65.34          45.03                31.08  
Handball                    92.80          69.59                25.01  
Judo                      

In [379]:
heigh_diff = male_vs_female_height['Height_diff_percent'].apply(lambda value: abs(value)).sort_values(ascending=False)
heigh_diff.describe()

count    33.000000
mean      6.559394
std       2.342812
min       2.290000
25%       5.170000
50%       6.110000
75%       8.200000
max      12.940000
Name: Height_diff_percent, dtype: float64

In [380]:
heigh_diff[heigh_diff > heigh_diff.describe()['75%']]

Sport
Alpine Skiing       12.94
Gymnastics          10.88
Triathlon            9.90
Tennis               9.69
Beach Volleyball     9.28
Bobsleigh            8.92
Archery              8.43
Swimming             8.41
Name: Height_diff_percent, dtype: float64

In [381]:
weigh_diff = male_vs_female_weight['Weight_diff_percent'].apply(lambda value: abs(value)).sort_values(ascending=False)
weigh_diff.describe()

count    33.000000
mean     22.182424
std       7.258357
min       1.510000
25%      17.910000
50%      23.400000
75%      26.000000
max      38.540000
Name: Weight_diff_percent, dtype: float64

In [382]:
# since std is high, difference is better show with mean
weigh_diff[weigh_diff > weigh_diff.describe()['mean']]

Sport
Alpine Skiing           38.54
Wrestling               31.47
Gymnastics              31.08
Weightlifting           30.28
Beach Volleyball        29.25
Rowing                  28.98
Cross Country Skiing    27.85
Triathlon               26.90
Golf                    26.00
Swimming                25.69
Modern Pentathlon       25.34
Rugby Sevens            25.30
Handball                25.01
Shooting                23.73
Archery                 23.60
Cycling                 23.54
Judo                    23.40
Equestrianism           22.50
Name: Weight_diff_percent, dtype: float64

#### <span id='pt1.4.5'></span>Qual foi (ou quais foram) o maior medalhista brasileiro em quantidade total de medalhas?

In [383]:
', '.join(df_brazil_medals['Name'].value_counts().index[:2])

'Robert Scheidt, Torben Schmidt Grael'

#### <span id='pt1.4.6'></span>E o(s) maior(es) em quantidade de medalhas de ouro?

In [384]:
df_brazil_medals[df_brazil_medals['Medal']=='Gold']['Name'].value_counts()

Marcelo Bastos Ferreira                               2
Adhemar Ferreira da Silva                             2
Sheilla Tavares de Castro Blassioli                   2
Jaqueline Maria "Jaque" Pereira de Carvalho Endres    2
Fabiana Marcelino Claudino                            2
                                                     ..
Neymar da Silva Santos Jnior                          1
Danielle "Dani" Rodrigues Lins                        1
Janelson dos Santos Carvalho                          1
Alexander Welter                                      1
Joaquim Carvalho Cruz                                 1
Name: Name, Length: 96, dtype: int64

#### <span id='pt1.4.7'></span>Qual esporte rendeu mais medalhas de ouro para o Brasil? E qual rendeu mais medalhas no total?

**DICA:** tome muito cuidado nessa análise: cada **evento esportivo** rende 1 medalha. Por exemplo, quando a equipe de futebol vence, isso é considerado 1 medalha, mesmo tendo cerca de 20 atletas medalhistas na equipe. 

In [385]:
df_brazil.columns

Index(['ID', 'Name', 'Sex', 'Age', 'Height', 'Weight', 'Team', 'NOC', 'Games',
       'Year', 'Season', 'City', 'Sport', 'Event', 'Medal'],
      dtype='object')

In [386]:
# select only the sports, events and by year
df_unique_by_sport = df_brazil[['Year','Sport','Event','Medal']].copy()
df_unique_by_sport.shape

(3828, 4)

In [387]:
# remove duplicates, like there were 1 athlete per sport
df_unique_by_sport.drop_duplicates(inplace=True)
df_unique_by_sport.shape

(1647, 4)

In [388]:
df_brazil_medals_by_sport = df_unique_by_sport.groupby(by='Sport')['Medal'].value_counts()
df_brazil_medals_by_sport.head()

Sport       Medal 
Athletics   Bronze    8
            Gold      5
            Silver    3
Basketball  Bronze    4
            Silver    1
Name: Medal, dtype: int64

In [389]:
# accessing the multindex dataframe
print('Number of gold medals by sport:')
df_brazil_medals_by_sport.loc[:,'Gold'].sort_values(ascending=False)

Number of gold medals by sport:


Sport
Sailing             7
Athletics           5
Volleyball          5
Judo                4
Beach Volleyball    3
Boxing              1
Equestrianism       1
Football            1
Gymnastics          1
Shooting            1
Swimming            1
Name: Medal, dtype: int64

In [390]:
print('Number of total medals by sport:')
df_brazil_medals_by_sport.groupby('Sport').sum().sort_values(ascending=False)

Number of total medals by sport:


Sport
Judo                 22
Sailing              18
Athletics            16
Swimming             14
Beach Volleyball     13
Volleyball           10
Football              8
Basketball            5
Boxing                5
Gymnastics            4
Shooting              4
Canoeing              3
Equestrianism         3
Taekwondo             2
Modern Pentathlon     1
Name: Medal, dtype: int64

#### <span id='pt1.4.8'></span>Cada "categoria" dentro de um esporte é considerado um evento. Por exemplo, dentro de "atletismo", temos uma competição de 100m masculina, uma de 100m feminino, um revezamento 4 x 100m masculino, um revezamento 4 x 100m feminino, uma competição de 400m masculino, uma de 400m feminino, uma maratona masculina, uma maratona feminina, e assim sucessivamente.

Sabendo disso, qual evento esportivo mais rendeu medalhas de ouro para o Brasil? E total de medalhas?

In [391]:
df_unique_by_event = df_brazil[['Year','Event','Medal']].copy()
df_unique_by_event.drop_duplicates(inplace=True)
df_brazil_medals_by_event = df_unique_by_event.groupby(by='Event')['Medal'].value_counts()
df_brazil_medals_by_event

Event                                 Medal 
Athletics Men's 200 metres            Bronze    1
Athletics Men's 4 x 100 metres Relay  Bronze    2
                                      Silver    1
Athletics Men's 800 metres            Gold      1
                                      Silver    1
                                               ..
Taekwondo Women's Heavyweight         Bronze    1
Volleyball Men's Volleyball           Gold      3
                                      Silver    3
Volleyball Women's Volleyball         Bronze    2
                                      Gold      2
Name: Medal, Length: 90, dtype: int64

In [392]:
print('Event with most gold medals: ',end='')
print(df_brazil_medals_by_event.loc[:,'Gold'].sort_values(ascending=False).index[0])

print('Event with most medals in total: ',end='')
print(df_brazil_medals_by_event.groupby('Event').sum().sort_values(ascending=False).index[0])

Event with most gold medals: Volleyball Men's Volleyball
Event with most medals in total: Beach Volleyball Women's Beach Volleyball


#### <span id='pt1.4.9'></span>Para finalizar sobre o Brasil: obtenha o total de medalhas de ouro, prata, bronze e total por ano.

In [393]:
medals_by_year = df_brazil_medals.groupby('Year')['Medal'].value_counts()
medals_by_year.head()

Year  Medal 
1920  Bronze     5
      Gold       1
      Silver     1
1948  Bronze    10
1952  Bronze     2
Name: Medal, dtype: int64

In [394]:
total_medals_by_year = medals_by_year.unstack()
total_medals_by_year.assign(Total=total_medals_by_year.sum(axis=1)).stack().to_frame('#')

#
Year Medal       
1920 Bronze   5.0
     Gold     1.0
     Silver   1.0
     Total    7.0
1948 Bronze  10.0
     Total   10.0
1952 Bronze   2.0
     Gold     1.0
     Total    3.0
1956 Gold     1.0
     Total    1.0
1960 Bronze  13.0
     Total   13.0
1964 Bronze  12.0
     Total   12.0
1968 Bronze   3.0
     Silver   1.0
     Total    4.0
1972 Bronze   2.0
     Total    2.0
1976 Bronze   3.0
     Total    3.0
1980 Bronze   5.0
     Gold     4.0
     Total    9.0
1984 Bronze   2.0
     Gold     1.0
     Silver  33.0
     Total   36.0
1988 Bronze   5.0
     Gold     1.0
     Silver  17.0
     Total   23.0
1992 Gold    13.0
     Silver   1.0
     Total   14.0
1996 Bronze  43.0
     Gold     5.0
     Silver  15.0
     Total   63.0
2000 Bronze  36.0
     Silver  12.0
     Total   48.0
2004 Bronze   3.0
     Gold    18.0
     Silver  19.0
     Total   40.0
2008 Bronze  30.0
     Gold    14.0
     Silver  34.0
     Total   78.0
2012 Bronze  11.0
     Gold    14.0
     Silver  34.0
     Total   59.0
2016 Bronze   6.0
     Gold    36.0
     Silver   8.0
     Total   50.0

## <span id='pt2' style='color:ForestGreen'><b>2. O mundo nos jogos de verão</b></span>

Vamos agora analisar um pouquinho do que aconteceu nas Olimpíadas de verão em todo o mundo.

### <span id='pt2.1' style='color:Gold'>Retome o DataFrame original e descarte as informações sobre os jogos de inverno.</span>

In [395]:
df_summer = df[df['Season']=='Summer']
df_summer.head()

ID                                Name Sex   Age  Height  Weight  \
0    1                           A Dijiang   M  24.0   180.0    80.0   
1    2                            A Lamusi   M  23.0   170.0    60.0   
2    3                 Gunnar Nielsen Aaby   M  24.0     NaN     NaN   
3    4                Edgar Lindenau Aabye   M  34.0     NaN     NaN   
26   8  Cornelia "Cor" Aalten (-Strannood)   F  18.0   168.0     NaN   

              Team  NOC        Games  Year  Season         City       Sport  \
0            China  CHN  1992 Summer  1992  Summer    Barcelona  Basketball   
1            China  CHN  2012 Summer  2012  Summer       London        Judo   
2          Denmark  DEN  1920 Summer  1920  Summer    Antwerpen    Football   
3   Denmark/Sweden  DEN  1900 Summer  1900  Summer        Paris  Tug-Of-War   
26     Netherlands  NED  1932 Summer  1932  Summer  Los Angeles   Athletics   

                           Event Medal  
0    Basketball Men's Basketball   NaN  
1   Judo Men's Extra-Lightweight   NaN  
2        Football Men's Football   NaN  
3    Tug-Of-War Men's Tug-Of-War  Gold  
26  Athletics Women's 100 metres   NaN

### <span id='pt2.2' style='color:Gold'>Obtenha a lista de todos os esportes já disputados nas olimpíadas de verão.</span>

In [396]:
df_summer['Sport'].unique()

array(['Basketball', 'Judo', 'Football', 'Tug-Of-War', 'Athletics',
       'Swimming', 'Badminton', 'Sailing', 'Gymnastics',
       'Art Competitions', 'Handball', 'Weightlifting', 'Wrestling',
       'Water Polo', 'Hockey', 'Rowing', 'Fencing', 'Equestrianism',
       'Shooting', 'Boxing', 'Taekwondo', 'Cycling', 'Diving', 'Canoeing',
       'Tennis', 'Modern Pentathlon', 'Golf', 'Softball', 'Archery',
       'Volleyball', 'Synchronized Swimming', 'Table Tennis', 'Baseball',
       'Rhythmic Gymnastics', 'Rugby Sevens', 'Trampolining',
       'Beach Volleyball', 'Triathlon', 'Rugby', 'Lacrosse', 'Polo',
       'Cricket', 'Ice Hockey', 'Racquets', 'Motorboating', 'Croquet',
       'Figure Skating', 'Jeu De Paume', 'Roque', 'Basque Pelota',
       'Alpinism', 'Aeronautics'], dtype=object)

### <span id='pt2.3' style='color:Gold'>Obtenha a lista de todas as modalidades esportivas já disputadas nas olimpíadas de verão.</span>

In [397]:
df_summer['Event'].unique()

array(["Basketball Men's Basketball", "Judo Men's Extra-Lightweight",
       "Football Men's Football", "Tug-Of-War Men's Tug-Of-War",
       "Athletics Women's 100 metres",
       "Athletics Women's 4 x 100 metres Relay",
       "Swimming Men's 400 metres Freestyle", "Badminton Men's Singles",
       "Sailing Women's Windsurfer",
       "Swimming Men's 200 metres Breaststroke",
       "Swimming Men's 400 metres Breaststroke",
       "Gymnastics Men's Individual All-Around",
       "Gymnastics Men's Team All-Around",
       "Gymnastics Men's Floor Exercise", "Gymnastics Men's Horse Vault",
       "Gymnastics Men's Parallel Bars",
       "Gymnastics Men's Horizontal Bar", "Gymnastics Men's Rings",
       "Gymnastics Men's Pommelled Horse", "Athletics Men's Shot Put",
       'Art Competitions Mixed Sculpturing, Unknown Event',
       "Handball Women's Handball",
       "Weightlifting Women's Super-Heavyweight",
       "Wrestling Men's Light-Heavyweight, Greco-Roman",
       "Gymnastics M

### <span id='pt2.4' style='color:Gold'>Obtenha a lista de todos os países que já disputaram olimpíadas.</span>

In [398]:
# wikipedia page where cotains information about NOC(National Olympic Committees) codes
url = 'https://en.wikipedia.org/wiki/List_of_IOC_country_codes'
# open the url with urlib.request tool
response = urlopen(url)
# read the response and parse into html
soup = bs(response.read(), 'html.parser')

In [399]:
# function for extracing the web scraping data
def extract_from_table(table, extract_all_columns=False):
    # get columns
    column_names = table.tbody('tr')[0].get_text().split('\n')
    # skip the empty value(tab) on the list, select the index 2 and 4
    column_names = column_names[1:4:2]
    # get rows
    # text slicing: skip the first index (paragraph) and select values only for the fist and second columns 
    rows = [text.get_text().split('\n')[1:3] for text in table.tbody('tr')[1:]]
    # code is in the first index of the list
    codes = [row[0] for row in rows]
    # country is in the second index of the list, and remove the special character from first index
    countries = [row[1][1:] for row in rows]
    # create pd data frame with the extracted values
    table_df = pd.DataFrame(data={
        column_names[0] : codes,
        column_names[1]: countries
    }, columns=column_names)

    return table_df

# Current NOCs table
current_noc_df = extract_from_table(table=soup('table')[0])
# Historic NOCs and teams: Codes still in use
old_codes_df = extract_from_table(table=soup('table')[2])
# Historic NOCs and teams: Special codes for Olympics
special_codes_df = extract_from_table(table=soup('table')[4])

In [400]:
# list of NOCs from the main data frame
print(f"Number of unique elements in NOC: {len(df['NOC'].unique())}")
print(f"Is there at least one entry in 'Codes still in use' table? {any(old_codes_df['Code'].isin(df['NOC'].unique()))}")
print(f"Is there at least one entry in 'Special Code' table? {any(special_codes_df['Code'].isin(df['NOC'].unique()))}")

Number of unique elements in NOC: 230
Is there at least one entry in 'Codes still in use' table? True
Is there at least one entry in 'Special Code' table? True


In [401]:
# create dataframe for the unique NOCs found in the main df
unique_noc = pd.DataFrame(data=df['NOC'].unique(), columns=['Code'])
# merge unique NOCs with each Code dataframe
unique_noc_df = unique_noc.merge(current_noc_df, how='left')
old_codes_df = unique_noc.merge(old_codes_df, how='left')
special_codes_df = unique_noc.merge(special_codes_df, how='left')
# fill NA values with the remaining dataframes 'Codes still in use' and 'Special Code'
unique_noc_df['National Olympic Committee'].fillna(old_codes_df['Nation/Team'], inplace=True)
unique_noc_df['National Olympic Committee'].fillna(special_codes_df['Nation/Team'], inplace=True)
unique_noc_df

Code National Olympic Committee
0    CHN                      China
1    DEN                    Denmark
2    NED                Netherlands
3    USA              United States
4    FIN                    Finland
..   ...                        ...
225  TUV                     Tuvalu
226  NFL                        NaN
227  KOS                     Kosovo
228  SSD                South Sudan
229  LES                    Lesotho

[230 rows x 2 columns]

In [402]:
print(unique_noc_df.isna().value_counts())
print('There is still some codes that doesnt have a matching country, these are from the Obsolete Codes: ')
print(list(unique_noc_df[unique_noc_df['National Olympic Committee'].isna()]['Code']))

Code   National Olympic Committee
False  False                         218
       True                           12
dtype: int64
There is still some codes that doesnt have a matching country, these are from the Obsolete Codes: 
['UAR', 'LIB', 'YAR', 'YMD', 'MAL', 'CRT', 'SAA', 'RHO', 'WIF', 'NBO', 'UNK', 'NFL']


In [403]:
# get 'Obsolete codes' table
columns = soup('table')[3].tbody('tr')[0].get_text().split('\n')[1::2]
rows = [text.get_text().split('\n')[1:5] for text in soup('table')[3].tbody('tr')[1:]]
obsolete_codes_df = pd.DataFrame(data={
    columns[0] : [row[0] for row in rows], #codes
    columns[1] : [row[1][1:] for row in rows], #nations
    columns[3] : [re.findall(r'\((.*?)\)', row[3]) for row in rows] #current noc
})
# correct empty list on merged table from the scrap data
obsolete_codes_df.iloc[8,2] = obsolete_codes_df.iloc[7,2]
obsolete_codes_df.iloc[18,2] = obsolete_codes_df.iloc[17,2]
# select only the first item
obsolete_codes_df['Notes'] = obsolete_codes_df['Notes'].apply(lambda x : x[0])
# merge the data frame with unique NOCs
obsolete_codes_df = unique_noc.merge(obsolete_codes_df, how='left')
obsolete_codes_df.dropna()

Code                                       Nation (NOC) Notes
36   UAR                               United Arab Republic   EGY
62   YAR           North Yemencode from Yemen Arab Republic   YEM
144  GUI                                     British Guiana   GUY
146  YMD     South Yemencode from Yemen Democratic Republic   YEM
168  MAL                       MalaysiaFrom French Malaisie   MAS
190  SAA                                               Saar   FRG
201  RHO  Rhodesiaalso Southern Rhodesia and Federation ...   ZIM
215  NBO                                       North Borneo   MAS

In [404]:
print(f"Check if all values from Notes(Current NOCs) are in Unique NOCs data frame: \
{obsolete_codes_df['Notes'].isin(unique_noc_df['Code']).value_counts()[1] == len(obsolete_codes_df.dropna())}")
# repeat the analysis and check if there is NaN values remaining
temp_df = unique_noc_df.copy()
temp_df['National Olympic Committee'].fillna(obsolete_codes_df['Nation (NOC)'], inplace=True)
print(f"Number of NaN values left: {temp_df['National Olympic Committee'].isna().value_counts()[1]}")
print(f"Codes with NaN values left: {list(temp_df[temp_df['National Olympic Committee'].isna()]['Code'])}")

Check if all values from Notes(Current NOCs) are in Unique NOCs data frame: True
Number of NaN values left: 5
Codes with NaN values left: ['LIB', 'CRT', 'WIF', 'UNK', 'NFL']


In [405]:
print('Names of the countries to participate in the Olympic games: ')
print(list(unique_noc_df['National Olympic Committee'].dropna().sort_values()))

Names of the countries to participate in the Olympic games: 
['Afghanistan', 'Albania', 'Algeria', 'American Samoa', 'Andorra', 'Angola', 'Antigua and Barbuda', 'Argentina', 'Armenia', 'Aruba', 'Australasia', 'Australia', 'Austria', 'Azerbaijan', 'Bahamas', 'Bahrain', 'Bangladesh', 'Barbados', 'Belarus', 'Belgium', 'Belize', 'Benin', 'Bermuda', 'Bhutan', 'Bohemia', 'Bolivia', 'Bosnia and Herzegovina', 'Botswana', 'Brazil', 'British Virgin Islands', 'Brunei', 'Bulgaria', 'Burkina Faso', 'Burundi', 'Cambodia', 'Cameroon', 'Canada', 'Cape Verde', 'Cayman Islands', 'Central African Republic', 'Chad', 'Chile', 'China', 'Chinese Taipei[7]', 'Colombia', 'Comoros', 'Cook Islands', 'Costa Rica', 'Croatia', 'Cuba', 'Cyprus', 'Czech Republic', 'Czechoslovakia', 'Democratic Republic of the Congo', 'Denmark', 'Djibouti', 'Dominica', 'Dominican Republic', 'East Germany', 'East Timor', 'Ecuador', 'Egypt', 'El Salvador', 'Equatorial Guinea', 'Eritrea', 'Estonia', 'Eswatini', 'Ethiopia', 'Federated Sta

### <span id='pt2.5' style='color:Gold'>Qual atleta foi o maior medalhista (em medalhas totais) da história das olimpíadas de verão?</span>

In [406]:
df_summer_medals = df_summer.dropna(subset=['Medal'])
df_summer_medals['Name'].value_counts().index[0]

'Michael Fred Phelps, II'

### <span id='pt2.6' style='color:Gold'>Qual atleta foi o maior medalhista de ouro da história das olimpíadas de verão?</span>

In [407]:
df_summer_medals[df_summer_medals['Medal']=='Gold']['Name'].value_counts().index[0]

'Michael Fred Phelps, II'

### <span id='pt2.7' style='color:Gold'>Qual país foi o maior medalhista de ouro da história das olimpíadas de verão? Lembre-se da questão do evento esportivo, para não considerar múltiplas medalhas para um mesmo evento (ex: uma equipe de futebol fazendo parecer que mais de 20 medalhas foram distribuídas).</span>

In [408]:
df_summer_medals_by_event = df_summer_medals[['Team','Year','Event','Medal']].copy()
df_summer_medals_by_event.drop_duplicates(inplace=True)
df_summer_medals_by_event.groupby(by='Team')['Medal'].value_counts().loc[:,'Gold'].sort_values(ascending=False).index[0]

'United States'

### <span id='pt2.8' style='color:Gold'>Qual país foi o maior medalhista em medalhas totais na história das olimpíadas de verão?</span>

In [409]:
df_summer_medals_by_event.groupby(by='Team')['Medal'].value_counts().groupby(by='Team').sum().sort_values(ascending=False).index[0]

'United States'

### <span id='pt2.9' style='color:Gold'>Obtenha o total de medalhas de ouro, prata e total por edição das Olimpíadas de verão. Lembre-se da questão do evento esportivo.</span>

In [410]:
summer_medals = df_summer_medals_by_event.groupby(by='Year')['Medal'].value_counts().unstack()
summer_medals.assign(Total=summer_medals.sum(axis=1)).stack().to_frame('#')

#
Year Medal      
1896 Bronze   34
     Gold     43
     Silver   41
     Total   118
1900 Bronze   92
...          ...
2012 Total   962
2016 Bronze  360
     Gold    307
     Silver  306
     Total   973

[116 rows x 1 columns]

## <span id='pt3' style='color:ForestGreen'><b>3. Brasil vs Mundo</b></span>

### <span id='pt3.1' style='color:Tomato'>Para finalizar, vamos fazer algumas comparações entre Brasil e mundo. Qual o ranking do Brasil em cada edição das olimpíadas? Lembrando que o ranking é ordenado por medalhas de ouro.</span>

In [411]:
# prepare data to remove duplicated medals and NAN values
df_medals = df[['NOC','Year','Event','Medal']].copy()
df_medals.dropna(inplace=True)
df_medals.drop_duplicates(inplace=True)

# get rankings by year
df_medals_by_year = df_medals.groupby(by=['Year','NOC'])['Medal'].value_counts().to_frame('Amount')
# transform index to columns
df_medals_by_year.reset_index(inplace=True)

# data frame for create the Ranking column
df_ranking = df_medals_by_year.copy()

# function to attribute score for the medals
def medal_to_rank(row):
    if row == 'Bronze':
        return 1
    elif row == 'Silver':
        return 100
    elif row == 'Gold':
        return 10000
# create a temp column for the total score of each medal
df_ranking['medal_score'] = df_ranking['Medal'].apply(medal_to_rank)*df_ranking['Amount']
# group data by Year
df_ranking_year = df_ranking.groupby('Year')
# for each group of Year, group each Team, sum its score and rank within the Team group
ranking = df_ranking_year.apply(lambda x : x.groupby('NOC')['medal_score'].sum().rank(method='dense',ascending=False))
# prepare data for merge turing into a data frame
ranking = ranking.reset_index()
ranking.rename(columns = {'medal_score':'Ranking'}, inplace=True)
ranking["Ranking"] = ranking["Ranking"].astype("int64")
# remove the temp column for the medal scores
df_ranking.drop('medal_score',axis=1,inplace=True)
# merge the two dataframes based on the Year and team groups
df_ranking = df.merge(ranking, how='inner', left_on=['Year','NOC'], right_on=['Year', 'NOC'])
df_ranking

ID                                Name Sex   Age  Height  Weight  \
0            1                           A Dijiang   M  24.0   180.0    80.0   
1         6848                      Bai Chongguang   M  21.0   184.0    83.0   
2         6854                             Bai Mei   F  17.0   166.0    46.0   
3        11225                            Bi Zhong   M  23.0   188.0   110.0   
4        17295                          Cai Yanshu   M  28.0   169.0    79.0   
...        ...                                 ...  ..   ...     ...     ...   
243597   97066  Norman Gilbert Pritchard (-Trevor)   M  24.0     NaN     NaN   
243598  119534                 Michel Johann Thato   M  22.0     NaN     NaN   
243599  125746                       Bart Veldkamp   M  30.0   181.0    84.0   
243600  125746                       Bart Veldkamp   M  30.0   181.0    84.0   
243601  125746                       Bart Veldkamp   M  30.0   181.0    84.0   

              Team  NOC        Games  Year  Season       City  \
0            China  CHN  1992 Summer  1992  Summer  Barcelona   
1            China  CHN  1992 Summer  1992  Summer  Barcelona   
2            China  CHN  1992 Summer  1992  Summer  Barcelona   
3            China  CHN  1992 Summer  1992  Summer  Barcelona   
4            China  CHN  1992 Summer  1992  Summer  Barcelona   
...            ...  ...          ...   ...     ...        ...   
243597       India  IND  1900 Summer  1900  Summer      Paris   
243598  Luxembourg  LUX  1900 Summer  1900  Summer      Paris   
243599     Belgium  BEL  1998 Winter  1998  Winter     Nagano   
243600     Belgium  BEL  1998 Winter  1998  Winter     Nagano   
243601     Belgium  BEL  1998 Winter  1998  Winter     Nagano   

                      Sport                                   Event   Medal  \
0                Basketball             Basketball Men's Basketball     NaN   
1                    Boxing          Boxing Men's Light-Heavyweight     NaN   
2       Rhythmic Gymnastics  Rhythmic Gymnastics Women's Individual     NaN   
3                 Athletics            Athletics Men's Hammer Throw     NaN   
4             Weightlifting   Weightlifting Men's Light-Heavyweight     NaN   
...                     ...                                     ...     ...   
243597            Athletics      Athletics Men's 200 metres Hurdles  Silver   
243598            Athletics                Athletics Men's Marathon    Gold   
243599        Speed Skating        Speed Skating Men's 1,500 metres     NaN   
243600        Speed Skating        Speed Skating Men's 5,000 metres  Bronze   
243601        Speed Skating       Speed Skating Men's 10,000 metres     NaN   

        Ranking  
0             4  
1             4  
2             4  
3             4  
4             4  
...         ...  
243597       17  
243598       14  
243599       20  
243600       20  
243601       20  

[243602 rows x 16 columns]

In [412]:
print('Brazil ranking by year:')
#print(df_ranking[['Year','Ranking']].loc[df_ranking['NOC']=='BRA'].reset_index(drop=True).drop_duplicates().to_string(index=False))
df_ranking[['Year','Ranking']].loc[df_ranking['NOC']=='BRA'].reset_index(drop=True).drop_duplicates().sort_values('Year').style.hide_index()

Brazil ranking by year:


### <span id='pt3.2' style='color:Gold'>Compare o maior medalhista em ouros do Brasil com o maior medalhista em ouros do mundo.</span>

In [413]:
print(f"Maior medalhista em ouro do mundo: {df[df['Medal']=='Gold']['Name'].value_counts().index[0]}")
br_gold_medalist = df.loc[(df['Medal']=='Gold') & (df['NOC']=='BRA')]['Name'].value_counts()
br_top_gold_medalist = br_gold_medalist[br_gold_medalist==max(br_gold_medalist)].index
print(f"Maior medalhista(s) em ouro do Brasil: {', '.join(br_top_gold_medalist)}")

Maior medalhista em ouro do mundo: Michael Fred Phelps, II
Maior medalhista(s) em ouro do Brasil: Thasa Daher de Menezes, Giovane Farinazzo Gvio, Adhemar Ferreira da Silva, Torben Schmidt Grael, Marcelo Bastos Ferreira, Srgio "Escadinha" Dutra dos Santos, Robert Scheidt, Maurcio Camargo Lima, Fabiana Marcelino Claudino, Sheilla Tavares de Castro Blassioli, Fabiana "Fabi" Alvim de Oliveira, Jaqueline Maria "Jaque" Pereira de Carvalho Endres, Paula Renata Marques Pequeno


### <span id='pt3.4' style='color:Gold'>Compare o maior medalhista em total de medalhas do Brasil com o maior medalhista em total de medalhas do mundo.</span>

In [414]:
df_medal = df.dropna()
print(f"Maior medalhista em medalha totais do mundo: {df_medal['Name'].value_counts().index[0]}")
br_total_medalist = df_medal.loc[df_medal['NOC']=='BRA']['Name'].value_counts()
br_top_total_medalist = br_total_medalist[br_total_medalist==max(br_total_medalist)].index
print(f"Maior medalhista(s) em medalha totais do Brasil: {', '.join(br_top_total_medalist)}")


Maior medalhista em medalha totais do mundo: Michael Fred Phelps, II
Maior medalhista(s) em medalha totais do Brasil: Robert Scheidt, Torben Schmidt Grael


### <span id='pt3.5' style='color:Gold'>Compare o maior medalhista em ouros do Brasil com o maior medalhista do mundo no mesmo esporte.</span>

In [415]:
sports = df_medal.loc[df_medal['Name'].isin(br_top_gold_medalist)]['Sport'].unique()
print(f"Esporte em que os maiores medalhistas de ouro do Brasil competiram: {', '.join(sports)}")
for sport in sports:
    print(f"Maior medalhista no esporte {sport}: ", end='')
    print(df_medal.loc[(df_medal['Sport']==sport) & (df_medal['Medal']=='Gold')]['Name'].value_counts().index[0])

Esporte em que os maiores medalhistas de ouro do Brasil competiram: Athletics, Volleyball, Sailing
Maior medalhista no esporte Athletics: Raymond Clarence "Ray" Ewry
Maior medalhista no esporte Volleyball: Regla Maritza Bell McKenzie
Maior medalhista no esporte Sailing: Paul Bert Elvstrm


### <span id='pt3.6' style='color:Gold'>Compare o maior medalhista em total de medalhas do Brasil com o maior medalhista do mundo no mesmo esporte.</span>

In [416]:
sport = df_medal.loc[df_medal['Name'].isin(br_top_total_medalist)]['Sport'].unique()[0]
df_medal.loc[df_medal['Sport']==sport]['Name'].value_counts().head()

Robert Scheidt                    5
Charles Benedict "Ben" Ainslie    5
Torben Schmidt Grael              5
Paul Bert Elvstrm                 4
Alessandra Sensini                4
Name: Name, dtype: int64

Neste caso, os dois atletas brasileiros com mais medalhas no total estão entre os maiores medalhistas desse esporte.

### <span id='pt3.7' style='color:Gold'>Calcule o percentual de medalhas de ouro, prata e bronze que o Brasil ganhou em cada olimpíada.</span>

In [417]:
df_medals_by_year.groupby(by=['Year','NOC'])['Amount'].sum()

df_total_medals = df_medals_by_year.groupby(by=['Year', 'Medal']).agg({'Amount': 'sum'})
#.groupby(level=0).apply(lambda x: 100 * x / float(x.sum()))
df_total_medals.rename(columns={'Amount':'Total'}, inplace=True)

br_total_medals_year = df_medals_by_year.loc[df_medals_by_year['NOC']=='BRA'].groupby(by=['Year', 'Medal']).agg({'Amount': 'sum'})
br_total_medals_year = br_total_medals_year.merge(df_total_medals, how='inner', left_on=['Year','Medal'], right_on=['Year', 'Medal'])
br_total_medals_year['Percent'] = round((br_total_medals_year['Amount'] * 100) / br_total_medals_year['Total'],3)
print("Medalhas do Brasil das olimpiadas:")
# dict for order medals in gold > silver > bronze
custom_dict = {
    'Bronze':0,
    'Silver':1,
    'Gold':2
}
# display in sorted order, first by Year and then by Medal with custom key
br_total_medals_year.groupby(by='Year', group_keys = False).apply(lambda x: x.sort_values(by = 'Medal', ascending=False, key=lambda x : x.map(custom_dict)))

Medalhas do Brasil das olimpiadas:


Amount  Total  Percent
Year Medal                         
1920 Gold         1    155    0.645
     Silver       1    152    0.658
     Bronze       1    139    0.719
1948 Bronze       1    170    0.588
1952 Gold         1    171    0.585
     Bronze       2    180    1.111
1956 Gold         1    177    0.565
1960 Bronze       2    187    1.070
1964 Bronze       1    203    0.493
1968 Silver       1    207    0.483
     Bronze       2    215    0.930
1972 Bronze       2    245    0.816
1976 Bronze       2    253    0.791
1980 Gold         2    242    0.826
     Bronze       2    261    0.766
1984 Gold         1    263    0.380
     Silver       5    257    1.946
     Bronze       2    282    0.709
1988 Gold         1    286    0.350
     Silver       2    280    0.714
     Bronze       3    310    0.968
1992 Gold         2    315    0.635
     Silver       1    314    0.318
1996 Gold         3    271    1.107
     Silver       3    273    1.099
     Bronze       9    297    3.030
2000 Silver       6    300    2.000
     Bronze       6    326    1.840
2004 Gold         5    301    1.661
     Silver       2    300    0.667
     Bronze       3    326    0.920
2008 Gold         3    302    0.993
     Silver       4    302    1.325
     Bronze       9    353    2.550
2012 Gold         3    302    0.993
     Silver       5    304    1.645
     Bronze       9    356    2.528
2016 Gold         7    307    2.280
     Silver       6    306    1.961
     Bronze       6    360    1.667